File: public_html/crm/odr/analyzer/daily-sales-analysis.js
(function (root) {
  'use strict';
  const A = root.OdrAnalyzer;
  if (!A) return;

  const WEEKDAYS = ['вс', 'пн', 'вт', 'ср', 'чт', 'пт', 'сб'];

  A.parseDailySheet = function (ws, location) {
    const days = [];
    const max = ws.rowCount || 400;
    for (let r = 3; r <= max; r++) {
      const dateRaw = A.cellVal(ws, r, 1);
      const d = A.parseDate(dateRaw);
      if (!d) continue;
      const fact = A.toNum(A.cellVal(ws, r, 4));
      if (fact == null) continue;
      const bonus1 = A.toNum(A.cellVal(ws, r, 10)) || 0;
      const bonus2 = A.toNum(A.cellVal(ws, r, 11)) || 0;
      const bonuses = bonus1 + bonus2;
      const guests = A.toNum(A.cellVal(ws, r, 14));
      const bar = A.toNum(A.cellVal(ws, r, 17));
      const kitchen = A.toNum(A.cellVal(ws, r, 18));
      const shisha = A.toNum(A.cellVal(ws, r, 20));
      const cash = (A.toNum(A.cellVal(ws, r, 6)) || 0) + (A.toNum(A.cellVal(ws, r, 7)) || 0);
      const card = (A.toNum(A.cellVal(ws, r, 8)) || 0) + (A.toNum(A.cellVal(ws, r, 9)) || 0);
      const netRevenue = fact - bonuses;
      days.push({
        date: d.toISOString().slice(0, 10),
        weekday: A.cellVal(ws, r, 2) || WEEKDAYS[d.getDay()],
        plan: A.toNum(A.cellVal(ws, r, 3)),
        fact,
        planPct: A.toNum(A.cellVal(ws, r, 5)),
        cash, card, bonuses,
        netRevenue,
        guests,
        avgCheck: A.toNum(A.cellVal(ws, r, 15)),
        avgCheckNet: guests ? netRevenue / guests : null,
        bar, kitchen, shisha,
        barShare: fact && bar != null ? bar / fact : null,
        kitchenShare: fact && kitchen != null ? kitchen / fact : null,
        shishaShare: fact && shisha != null ? shisha / fact : null,
      });
    }

    const byWeekday = {};
    WEEKDAYS.forEach((w) => { byWeekday[w] = { revenue: 0, guests: 0, count: 0, avgCheck: 0 }; });
    days.forEach((day) => {
      const w = day.weekday || WEEKDAYS[new Date(day.date).getDay()];
      if (!byWeekday[w]) byWeekday[w] = { revenue: 0, guests: 0, count: 0 };
      byWeekday[w].revenue += day.netRevenue || 0;
      byWeekday[w].guests += day.guests || 0;
      byWeekday[w].count += 1;
    });
    Object.keys(byWeekday).forEach((w) => {
      const b = byWeekday[w];
      b.avgRevenue = b.count ? b.revenue / b.count : 0;
      b.avgGuests = b.count ? b.guests / b.count : 0;
      b.avgCheck = b.guests ? b.revenue / b.guests : 0;
    });

    const sorted = [...days].sort((a, b) => (b.netRevenue || 0) - (a.netRevenue || 0));
    const totalNet = days.reduce((s, d) => s + (d.netRevenue || 0), 0);
    const totalPlan = days.reduce((s, d) => s + (d.plan || 0), 0);
    const totalGuests = days.reduce((s, d) => s + (d.guests || 0), 0);

    const avgDaily = days.length ? totalNet / days.length : 0;
    const avgGuestsDay = days.length ? totalGuests / days.length : 0;

    return {
      location,
      days,
      stats: {
        daysCount: days.length,
        totalNet,
        totalPlan,
        planAchievement: totalPlan ? totalNet / totalPlan : null,
        totalGuests,
        avgDailyRevenue: avgDaily,
        avgDailyGuests: avgGuestsDay,
        avgCheck: totalGuests ? totalNet / totalGuests : null,
        avgKitchenShare: days.length ? days.reduce((s, d) => s + (d.kitchenShare || 0), 0) / days.length : null,
      },
      byWeekday,
      topBest: sorted.slice(0, 5),
      topWorst: sorted.slice(-5).reverse(),
      weakDays: days.filter((d) =>
        (d.netRevenue != null && d.netRevenue < avgDaily * 0.7)
        || (d.guests != null && d.guests < avgGuestsDay * 0.7)
        || (d.kitchenShare != null && d.kitchenShare < 0.18)
      ).slice(0, 10),
    };
  };
})(typeof window !== 'undefined' ? window : globalThis);

File: tools/odr-analyzer/daily_sales_analysis.py
"""Daily revenue sheets АДМИР / УДЕЛКА."""
from __future__ import annotations

from datetime import datetime, date
from typing import Any


def _num(v: Any) -> float | None:
    if v is None:
        return None
    try:
        return float(v)
    except (TypeError, ValueError):
        return None


def parse_daily(ws, location: str) -> dict:
    days = []
    for r in range(3, ws.max_row + 1):
        raw = ws.cell(r, 1).value
        if not isinstance(raw, (datetime, date)):
            continue
        fact = _num(ws.cell(r, 4).value)
        if fact is None:
            continue
        bonus = (_num(ws.cell(r, 10).value) or 0) + (_num(ws.cell(r, 11).value) or 0)
        net = fact - bonus
        guests = _num(ws.cell(r, 14).value)
        days.append({
            'date': raw.strftime('%Y-%m-%d') if hasattr(raw, 'strftime') else str(raw)[:10],
            'weekday': ws.cell(r, 2).value,
            'fact': fact, 'netRevenue': net, 'guests': guests,
            'bar': _num(ws.cell(r, 17).value), 'kitchen': _num(ws.cell(r, 18).value),
            'shisha': _num(ws.cell(r, 20).value),
            'kitchenShare': (_num(ws.cell(r, 18).value) / fact if fact and _num(ws.cell(r, 18).value) else None),
        })
    total_net = sum(d['netRevenue'] for d in days)
    sorted_days = sorted(days, key=lambda d: d['netRevenue'], reverse=True)
    return {
        'location': location, 'days': days,
        'stats': {'daysCount': len(days), 'totalNet': total_net, 'totalGuests': sum(d['guests'] or 0 for d in days)},
        'topBest': sorted_days[:5], 'topWorst': list(reversed(sorted_days[-5:])),
    }
